### PR TITLE
Fixed GPI operator to be differentiable and work with all interfaces

### PR DESCRIPTION
**Pull request by Gabriel-Bottrill**
Previously GPI operator didn't work with some interfaces and one couldn't retreive a gradient from it
Now GPI operator can be used with all interfaces and all interfaces can be used  to find gradient except for None.

All tests are passing and tests have been added for change.

Interfaces:

- "autograd"
- "pytorch"
- "tensorflow"
- "jax"
- None

### CODE REVIEW

File: ionizer/ops.py [hunk 1]
```diff
@@ -42,6 +45,8 @@ class GPI(Operation):
     num_wires = 1
     num_params = 1
     ndim_params = (0,)
+    grad_method = "A"
+    parameter_frequencies = [(1,)]
```
Comment (glassnotes) [inline ionizer/ops.py:50]: Please check into the documentation of the parameter_frequencies - it has to do with the eigenvalues of the generator, and is not a fixed value for all single-qubit operations. It's important because I believe the parameter-shift rule generator uses the frequencies to determine what the shift values and gradient recipe should be.

Comment (glassnotes) [reply ionizer/ops.py:50]: Please also define the `generator` method for the operation.

File: ionizer/ops.py [hunk 2]
```diff
@@ -62,7 +67,15 @@ def compute_matrix(phi):  # pylint: disable=arguments-differ
         array([[0.        +0.j        , 0.95533649-0.29552021j],
                [0.95533649+0.29552021j, 0.        +0.j        ]])
         """
-        return qml.math.stack([[0, qml.math.exp(-1j * phi)], [qml.math.exp(1j * phi), 0]])
+        if qml.math.get_interface(phi) == "tensorflow":
+                phi = qml.math.cast_like(phi, 1j)
+
+        a = (0 + 0j)
+        b = qml.math.exp((0 - 1j) * phi)
+        c = qml.math.exp((0 + 1j) * phi)
```
Comment (glassnotes) [inline ionizer/ops.py]: Instead of defining `c`, can use `qml.math.conj` in the matrix construction.

File: tests/test_ops.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+import pennylane as qml
+from pennylane import numpy as np
+import pennylane.math as math
+import tensorflow as tf
+from ionizer.ops import GPI
+
+jax.config.update("jax_enable_x64", True)
+
+interfaces = [
```
Comment (glassnotes) [inline tests/test_ops.py]: To avoid typing everything out again (and since you two are sitting right next to each other :laughing: ) please see the comments I've left on [the GPI2 PR](https://github.com/QSAR-UBC/ionizer/pull/2/files) regarding how to structure the interface testing, as well as trying with multiple diff methods.

File: tests/test_ops.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    def circuit(phi): 
+        qml.Hadamard(wires=0)
+        GPI(phi, wires=0)
+        return qml.expval(qml.PauliZ(wires=0))
+    
+    @staticmethod
+    def circuit_unitary(phi): 
+        qml.Hadamard(wires=0)
+        gpi_matrix = get_GPI_matrix(phi)
+        qml.QubitUnitary(gpi_matrix, wires=0)
```
Comment (glassnotes) [inline tests/test_ops.py]: This circuit and the test that uses it are testing the function defined in the test file rather than a library function :sweat_smile:

File: tests/test_ops.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            case "torch":
+                return torch.tensor(x)
+            case "tf":
+                return tf.Variable(x)
+            case _:
+                return qml.numpy.array(x)
+            
+    @pytest.mark.parametrize("interface", interfaces)
+    def test_GPI_compute_matrix(self, interface):
+        phi_rand = np.random.rand() * two_pi
```
Comment (glassnotes) [inline tests/test_ops.py:70]: Instead of `rand` can just pick a couple arbitrary values and use a `parametrize` decorator.

File: tests/test_ops.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        val_unitary = qnode_unitary(phi)
+        val_GPI = qnode_GPI(phi_GPI)
+
+        assert np.isclose(val_GPI, val_unitary)
+
+
+    @pytest.mark.parametrize("interface", interfaces)
+    def test_GPI_grad(self, interface):
+        phi = np.random.rand() * two_pi
```
Comment (glassnotes) [inline tests/test_ops.py:101]: Same comment as above re. using `rand`

File: tests/test_ops.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        phi_GPI = self.interface_array(phi, interface)
+        dev = qml.device("default.qubit", wires=1)
+
+        qnode_GPI = qml.QNode(self.circuit, dev, interface=interface)
+        qnode_unitary = qml.QNode(self.circuit_unitary, dev)
+
+        grad_unitary = qml.grad(qnode_unitary, argnum=0)(phi)
+
+        match interface:
```
Comment (glassnotes) [inline tests/test_ops.py:107]: Oh neat, I hadn't seen a `match` statement used in the wild yet!

**Review by glassnotes — CHANGES_REQUESTED** [review]
Thanks @Gabriel-Bottrill !

Many comments overlap with #2 so please refer to those as well and discuss amongst yourselves.

Also before re-uploading, could you please also format your code using black, for consistency?

```
black -l 100 file.py
```

File: tests/test_ops.py (hunk 6 above)
Comment (Gabriel-Bottrill) [reply tests/test_ops.py:101]: Should be fixed

File: tests/test_ops.py (hunk 5 above)
Comment (Gabriel-Bottrill) [reply tests/test_ops.py:70]: Should be fixed

File: tests/test_ops.py (hunk 4 above)
Comment (Gabriel-Bottrill) [reply tests/test_ops.py]: Should be fixed

File: ionizer/ops.py (hunk 2 above)
Comment (Gabriel-Bottrill) [reply ionizer/ops.py]: Fixed.
